### DISCUSSION TITLE
Modbus RTU lower than specification recommendation

### DISCUSSION THREAD

**Discussion by epneo-sebastianbaier** [zephyrproject-rtos/zephyr#78185] [Q&A]
Is the resulting Modbus RTU timeout not different from the recommended value from the specification?

https://github.com/zephyrproject-rtos/zephyr/blob/8c32a82e47f5d1a2871ec99b0c8b61eb789f9fcb/subsys/modbus/modbus_serial.c#L587-L592

This will result in a RTU timeout of ~1000 uS at 38400 bits/s, but the specification recommended a higher value:

> Remark :
The implementation of RTU reception driver may imply the management of a lot of interruptions due to the t1.5 and t3.5 timers. With high communication baud rates, this leads to a heavy CPU load. Consequently these two timers must be strictly respected when the baud rate is equal or lower than 19200 Bps. For baud rates greater than 19200 Bps, fixed values for the 2 timers should be used: it is recommended to use a value of 750µs for the inter-character time-out (t1.5) and a value of 1.750ms for inter-frame delay (t3.5). 

[Modbus Serial Line Protocol and Implementation Guide V1.02](https://modbus.org/docs/Modbus_over_serial_line_V1_02.pdf)

Does a lower value for the inter-frame delay of 1000 us not lead to an incompatibility between devices, which implement the recommend value of 1750 us?

**Comment by epneo-sebastianbaier** [comment]
Sorry @jfischer-no, for not initially creating a discussion about it.

I thought about whether this would lead to incompatibility between Modbus devices. In my opinion it should not, because in the remark from the specification it is about the "RTU receive driver" and a lower inter-frame time-out should allow the receiver to receive messages faster.

However, I have another point (more of a follow-up) that would be nice to have clarified:

The referenced source code limits the inter-frame timeout to ~1000 us (the calculated timeout at 38400 bits per second), but why? If the decision has been made not to use the recommended timeout of 1750 us, why then choose 1000 us instead of always calculating it based on the baud rate and just limiting it to the highest technically possible baud rate? 

It would be nice if we could clarify this topic. I think this is a problem when the interframe timeout is just randomly chosen.

**Comment by epneo-sebastianbaier** [comment]
No one here, who can say something about it?

**Reply by ndrs-pst** [reply]
Hi @epneo-sebastianbaier,

It's true that the value differs from the Modbus recommendation.

I suspect that 1,000 μs was selected to reduce the inter-frame timeout while still being large enough to distinguish the inter-character timeout.

Since Modbus only has a single master on the bus, I don't foresee any issues arising.

**Comment by jfischer-no** [comment]
@epneo-sebastianbaier Sorry, still have not had time to look why I preferred to limit it at 38400 bits/s
